notifications/models.py

```diff
@@ -19,38 +49,280 @@ class Subscription(models.Model):
     #    5) BillActionSubscription: actions on individual bills
     #    6) EventsSubscription: all events (e.g. https://nyc.councilmatic.org/events/ )
 
-    user = models.ForeignKey(User, related_name='%(class)ss', db_column='user_id')
-    last_datetime_updated = models.DateTimeField(auto_now=True) #XXX
+    user = models.ForeignKey(
+        User,
+        related_name='%(class)ss',
+        db_column='user_id',
+        on_delete=models.CASCADE
+    )
+    last_datetime_updated = models.DateTimeField(auto_now=True)
 
     # Make this an abstract base class
     class Meta:
         abstract = True
-    pass
+
+    def set_last_datetime_updated(self):
+        # Set the last_datetime_updated value to the current time, even if the
+        # model itself hasn't been updated. This is useful for knowing when the
+        # last notification went out, even if the subscription data itself hasn't
+        # changed.
+        self.last_datetime_updated = timezone.now()
+        self.save()
 
 
 class PersonSubscription(Subscription):
     # related_name lets us go from the user to their committee member subscriptions
-    person = models.ForeignKey(Person, related_name = 'subscriptions')
+    person = models.ForeignKey(
+        councilmatic_models.Person,
+        related_name='subscriptions',
+        on_delete=models.CASCADE
+    )
+    # Keep track of sponsorships that have already been seen, as an array of
+    # Bill.ids. This way, we can tell if sponsorships have been added or removed
+    # for this person.
+    seen_sponsorship_ids = ArrayField(OCDIDField(ocd_type='bill'), default=list)
+
+    def get_updates(self):
+        person_update = {}
+        new_sponsorships = councilmatic_models.BillSponsorship.objects.filter(
+            person__id=self.person.id
+        ).exclude(id__in=self.seen_sponsorship_ids)
+        if new_sponsorships.count() > 0:
+            person_update = {
+                'New sponsorships': {
+                    'slug': self.person.slug,
+                    'name': self.person.name,
+                    'bills': []
+                }
+            }
+            for sponsorship in new_sponsorships:
+                person_update['New sponsorships']['bills'].append({
+                    'id': sponsorship.id,
+                    'identifier': sponsorship.bill.identifier,
+                    'slug': sponsorship.bill.slug,
+                    'description': sponsorship.bill.title,
+                })
+        self.seen_sponsorship_ids.extend([spon.id for spon in new_sponsorships])
```

Comment:
do we need to be concerned about hitting an upper limit on array size in postgres, or about this array growing big enough to affect performance in the app?

Reply:
Good question! Looks like there's no size limit on Postgres Arrays, but there is a global size limit of 1GB for any individual field ([source](https://stackoverflow.com/a/29766223/7781189)). I don't think we'll hit that any time soon for Councilmatic, particularly in the narrow domain of Sponsorships tied to a specific person.

The performance question is a good one, and I really don't have a good sense of how to answer judge performance for this approach. The only other way I could think to accomplish this functionality would be to use a `ManyToManyField` to store `seen_sponsorships` but that seems to me like it would incur the same performance costs as this approach (namely having to retrieve all of the Sponsorships to get a list of IDs to exclude from the query above). Do you have any other ideas for how we could accomplish it that might be more performant?

Reply:
i think this is probably ok. the performance concern i had was mainly loading a big array into memory for the filter, but we'd have the same issue with a manytomanyfield, except we'd also need to worry about prefetching to keep the number of db queries low.

ty for checking on postgres! i think you're right that 1gb is a long way off... ed burke has 5400 sponsorships. i wrote a file with that many uuids to my desktop, and it's only 259 KB, hehe.